### ISSUE TITLE
Move to using Roam for content

### ISSUE THREAD

**Issue by necopinus** [vgururao/yakcollective#46]
Based on the conversation here and on Discord, move to using Roam for storing content. We _could_ wait for an API, but there's a desire to backup Roam periodically as well. If we just check those dumps into the GitHub repo, we can perhaps kill two birds with one stone.

---

> **Original ticket follows (so that the conversation makes sense)...**

I'm starting to think that content should be modularized and abstracted away from the pages themselves.

So, we have bits of content. I think these should still be markdown + HTML pages with auxiliary assets (PDFs, images, etc.).

Then, individual pages should be something like:

```xml
<Page settings="stuff" metadata="more stuff">
  <Page content="foo.md" presentation="block style 1" />
  <Page content="bar.html" presentation="block style 2" />
  <Page content="baz.md" presentation="block style 3" />
</Page>
```

Maybe content could then be maintained not directly through the admin interface, but via something like [Foam](https://foambubble.github.io/foam/)? (How do we handle cross-block links in the content, when a given linked block may be presented in multiple ways on the website?)

I think that this 100% requires migrating to Gatsby (#12).

**Comment by necopinus** [comment]
Marking this as a v3.0 thing for now, but this may be _way_ too ambitious...

**Comment by necopinus** [comment]
Marked as "help wanted", because I think there needs to be more thought put into how to do this, and folks' ideas, prior art, etc. would be _really_ helpful...

**Comment by necopinus** [comment]
_Tentative_ idea...

**(1)** Store content in a [Foam](https://foambubble.github.io/foam/) repo.

**(2)** Extend syntax a bit like the following:

```markdown
---
title: Page Title
date: 2020-07-18
embed: false
---

<!-- begin:include -->
[[foo]]
<!-- end:include -->

<!-- begin:include:a -->
[[bar]]
<!-- end:layout -->

<!-- begin:layout:b -->
[[baz]]
<!-- end:layout -->

[//begin]: # "Autogenerated link references for markdown compatibility"
[foo]: foo "foo"
[bar]: bar "bar"
[baz]: baz "baz"
[//end]: # "Autogenerated link references"
```

This file renders fine in Foam.

**(3)** Create a pre-processor that walks through data directory pre-build in Gatsby, replacing

```markdown
<!-- begin:layout:b -->
[[baz]]
<!-- end:layout -->
```

with

```xml
<Embed content="baz" layout="b" />
```

_Need to watch out for infinite recursion here!_

We may also want to do something in the pre-processor where we copy all assets into Gatsby's `static` directory (so that links continue to work.

**(4)** In Gatsby, use a custom component to render and embed the contents of `baz.md`, wrapped in `<div class="b"></div>`.

**(5)** Walk through all markdown files in the data directory, generating pages whenever `embed` is `false` (or missing).

---

This process should give us an editable, Roam-like data directory with the option to embed files in each other (allowing that file content to be used in multiple locations). Moreover, we can now create stacks of arbitrary layouts for our pages, allowing for a bit more flexibility between projects, etc.

**Pros:**

- More Roam-like experience for editing/navigation.
- More flexibility with page construction (layouts are relatively unimportant).
- Content is modularized and can be re-used.

**Cons:**

- Basically obsoletes the admin interface.
- Right now Foam subdirectory support works, but is a bit wonky (files are always created in the project root, and wikilinks only match the file name, not subdirectory + filename).

**Unknowns:**

- Does this raise the bar for contributing (because now there's no admin interface), or lower it (as now all content is represented in a more-or-less uniform way)?
- Internal links wind up looking like "[[Link](https://github.com/The-Yak-Collective/yakcollective/issues/46)]". Correct naming also depends on CSS tricks. Might be fine. Might even be kinda cool from a branding point of view (plain text FTW!). Might be fragile/off-putting.

**Comment by tilnow** [comment]
it seems like you want each project to also create and maintain their subset of webpages. i think this means using as standard an interface as possible. even a standard wiki may be too esoteric.

if this is not what you want, it may be simpler to define a small number of structures and hard-code them, so the end-users do not see too flexible a set of options.

**Comment by tilnow** [comment]
another consideration - roam is a few months from publishing an API (so they say). if/when they have one, it may be simpler to get this functionality from Roam itself, no?

**Comment by necopinus** [comment]
I go back-and-forth on the Roam thing. I know that everyone's super-into Roam, but it doesn't really use a flavor of Markdown that works well in other places, and I'm super-paranoid about lock-in and history. That said, I recently discovered [`gatsby-digital-garden`](https://github.com/mathieudutour/gatsby-digital-garden), which might be the starting point for something like this.

Foam is _mostly_ Roam-like, but because it's just bootstrapping off of markdown files it's missing the concept of "blocks". So I'm just spitballing here how we can get something like that functionality back. Foam's just a collection of VSCode extensions, so all you need to work with it is VSCode and a Git client (VSCode's even smart enough to auto-configure itself -- try cloning the [`foam-template`](https://github.com/foambubble/foam-template) repo and then opening the folder in VSCode).

What I _really_ want is Roam(-like) but backed by our own GitHub repo (so that we control the data and aren't dependent on their infrastructure), but I don't think such a thing exists (yet).

**Comment by tilnow** [comment]
what i don't like about foam is the need to use vscode. and the fidgety syntax 
i like the digital garden. it means you can always leave, no?
but there is no "push", so website is only for viewing. i assume each page could have a link to raom page, which mostly solves that.

i am starting to see why you like gatsby. FWIW, i do not mind building thinks for the jekyll back end and then rewriting for gatsby. i consider it "plan to throw one away, you will anyway"

**Comment by necopinus** [comment]
I'm curious what you find fidgety about Foam's syntax (it's just GitHub Flavored Markdown). VS Code isn't a hard dependency --- I use that on my laptop, but on my iPad I use Texstatic, and on my phone I use GitJournal. Both are more limited interfaces, and don't automatically generate the wiki-compatible links, but using them doesn't break anything. Anyways, this seems like a conversation we should have on Discord, rather than in this ticket.

The `gatsby-digital-garden` approach seems to be doing a JSON backup of Roam, and then loading the resulting files into a data structure. I remember seeing a different Gatsby project that loaded Roam dumps as well as its data source, though the URL/name is now escaping me.

There's less lock-in here, though it's still not great.

1. Roam content is pulled at build time, which will slow things down, and doesn't actually provide us with a record. I think a way around this would be to write a Netlify function that pulls a Roam backup and then commits it to GitHub, triggering a rebuild. Then we just use a periodic timer to trigger that function, say, once a day (and we can trigger it manually if needed)

2. If Roam ever disappears, we're still left with a JSON blob. We can parse that out into markdown files, but it's not something we really want to work on directly. My inclination is to keep data as legible as possible (modulo performance constraints, etc.).

We could live with (2), though I'll grumble about it for a while.

**Comment by necopinus** [comment]
Based on conversation in the #infrastructure channel, I'm changing this to using Roam backups as our data source.

**Comment by tilnow** [comment]
so lets get version 3 testbed up and running so we can play with this

**Comment by tilnow** [comment]
> I'm curious what you find fidgety about Foam's syntax (it's just GitHub Flavored Markdown). VS Code isn't a hard dependency --- I use that on my laptop, but on my iPad I use Texstatic, and on my phone I use GitJournal. Both are more limited interfaces, and don't automatically generate the wiki-compatible links, but using them doesn't break anything. Anyways, this seems like a conversation we should have on Discord, rather than in this ticket.
👍 
> 
> The `gatsby-digital-garden` approach seems to be doing a JSON backup of Roam, and then loading the resulting files into a data structure. I remember seeing a different Gatsby project that loaded Roam dumps as well as its data source, though the URL/name is now escaping me.
> 
> There's less lock-in here, though it's still not great.
> 
> 1. Roam content is pulled at build time, which will slow things down, and doesn't actually provide us with a record. I think a way around this would be to write a Netlify function that pulls a Roam backup and then commits it to GitHub, triggering a rebuild. Then we just use a periodic timer to trigger that function, say, once a day (and we can trigger it manually if needed)
it works by logging into roam. so yes.
> 2. If Roam ever disappears, we're still left with a JSON blob. We can parse that out into markdown files, but it's not something we really want to work on directly. My inclination is to keep data as legible as possible (modulo performance constraints, etc.).
> 
> We could live with (2), though I'll grumble about it for a while.
*i think that if raom disappears there will be more tools like gatsby-digital-garden. so i am not worried

**Comment by tilnow** [comment]
BTW, using the jekyll "{% include %}" directive we can also do build-time assembly. i guess the question is how sophisticated you want the assembly to be

**Comment by necopinus** [comment]
Roam export structure: https://davidbieber.com/snippets/2020-04-25-roam-json-export/

**Comment by tilnow** [comment]
want me to try this with jekyll? meaning - build web pages from the json backup?
give me  a sample backup please

for longer term use, we need to somehow get tese backup regularly. puppeteer (him from digital gardens) does that, but needs a password to get in.

**Comment by necopinus** [comment]
Not right now. There's questions about how we want to set up structure, etc. This was just me making a note for the future.

**Comment by tilnow** [comment]
Ok. Because this links in with question of proposal management. If we do them in roam rather than GitHub (as I suggested), then I should build some sort of demo of that.

**Comment by necopinus** [comment]
I agree that this interacts with the proposal management question... But I also worry that we've got too many irons in the fire right now. I think if you want to play around with something in Jekyll, let's do that on the side right now (PoC, rather than something built off of this repo).

I think structure is going to be the hard part. My initial thought was that we could use Roam namespaces to indicate which pages should be built, and how we should handle them. So, pages named `website/projects/Don't Waste the Reboot`, `website/members/Venkatesh Rao`, etc. But I've also thought about this all of 30 minutes so far.

FYI, I'm going to be at HOPE 2020 this next week, and working on things a lot less.

**Comment by tilnow** [comment]
enjoy HOPE. 
what do you mean by PoC?
anyway, i was going to work on the technical side - what does a page taken form roam look like. i am assuming that as long as we use roam as a tool for managing, the desire to at least publish one page of roam on a website will still be there. and i am wondering how much work it is to reproduce such a page in a usable way. simplest is to simply scrape  the html and show it. but is that sufficent?

**Comment by necopinus** [comment]
PoC = Proof of concept

My inclination is to stay away from web automation and scraping HTML off of web pages if there are other options. I have a few years of experience in both realms under my belt from the $dayjob, and I 100% guarantee that anything we build using those approaches is going to be _really_ fragile.

I'd start by seeing if you can get a JSON backup of the current ArtOfGig graph from @vgururao. [David Bieber's work](https://davidbieber.com/snippets/2020-04-25-roam-json-export/) (also linked above) look like a good starting point for reconstructing pages.

**Comment by tilnow** [comment]
1. scraping is quite fragile
2. they already have an export function. attached. what i am missing is some indication of how we want it translated into webpages. a page per page?
3. because they have a built-in  export, i expect any scraping we do (and none needed, i think) will be quickly superseded by using the API the keep on promising.
[Roam-Export-1595829477787.zip](https://github.com/The-Yak-Collective/yakcollective/files/4980184/Roam-Export-1595829477787.zip)

**Comment by tilnow** [comment]
meanwhile, i built a copy of the roam pages in our website: https://inspiring-villani-8898ca.netlify.app/roam/index/. the links work (or at least the ones i tested).
this doubles the build time, form 1 -> 2 minutes (ok, more than that in real life minutes)

also, the original roam pages do NOT have a unique url. if you want, though, each "component" can be linked back to the roam page it came form.

**Comment by necopinus** [comment]
This is... Basically what I'm thinking about.

I still think we should use namespaces in Roam, and it's missing some children, and formatting needs to be cleaned up... But those are all details.

This potentially removes one of the big pulls for moving to Gatsby. The main thing that's left is that Gatsby lets us easily use MDX (an extension of Markdown that includes React components), which should make it easier to create more modular/featureful pages.

I think we should check-in about the directionality here. I'll DM you on Discord, and we can set up a date.

**Comment by tilnow** [comment]
which brings up back to the question "what is the job of the webpage in the greater scheme of things". specifically, do we want more exciting pages? i discussed last night with jenna dixon the possibility of simply using iframes, whereby you can build featureful pages on some other tool (say a project page) and then embed it via iframe in website.

**Comment by vgururao** [comment]
The job of the webpage is generally to capture attention to projects via
the main call to action of attracting gigs to individual contributors and
YC as a whole for team projects. But it’s okay to get there inefficiently
via interestingness. Secondary function is to help new members orient
around what YC is and learn ways they can participate and benefit.

**Comment by necopinus** [comment]
I no longer think this is the way we want to go.
